src/common/stringutils.cpp

```diff
@@ -182,10 +182,6 @@ int StrUtils::Utf8CharSizeAt(const std::string &str, unsigned int pos)
     if((c & 0xE0) == 0xC0)
         return 2;
 
-    // Invalid char - unexpected continuation byte
```

Comment:
Why is this exception removed? It caught more than one bug in our UTF-8 handling IIRC so I think it should stay, otherwise bugs will hide more easily.

Reply:
I think more true variant is:
  
    if((c & 0x80) != 0) then exception
or

    if(( c & 0b10000000) != 0b00000000) then exception.

When char have size = 1 byte it must have mask 0b0xxxxxxx if byte does not match this mask, that means an error.

Reply:
I remember I needed to fix something in this function, and I consulted other team members regarding it, so I'm pretty sure it's correct. If the condition needs changing then so be it, but it's still an error and so there should be an exception.

Reply:
Ah, I only now saw the commit message, my bad.

> Show questions instead of crushing in case of UTF-8 errors 

I am not entirely sure that removing this exception is the right way to achieve that. It was there in the first place for a reason. If you want that behaviour then I think a better solution would be to catch this exception in places where the error is ok to be ignored which I assume is not everywhere where this function is used.

I'd rather let the game crash with the exception than let some subtle bugs hide away by removing this exception. See relevant PR https://github.com/colobot/colobot/pull/1276

Reply:
I recalled what the function does and now I am definitely against removing these lines of code.

> I think more true variant is:

Your condition is incorrect. The condition checks whether `c` is the continuation byte which is defined as a byte which starts with `10`. If `c` starts with bits `11` then `c & 0x80` would result in `11000000` which is definitely not `0` so your condition is true but `c` is *not* a continuation byte in this case.

The function is correct. It assumes it is given the first byte of a UTF-8 character and returns the size in bytes of the UTF-8 character. The first few bits of the first byte define how long is the character. All the other bytes belonging to the UTF-8 character must be continuation bytes. See https://stackoverflow.com/a/9356203

The exception is thrown in case the first byte is a continuation byte because it is simply wrong according to UTF-8 specification. So it means there's something wrong with the whole input text (it contains invalid UTF-8 characters) or the function was given a wrong position. In both cases, the fault lies higher in the stack and so the decision of what to do with this error should be decided outside this function, somewhere higher in the stack. And this can be done by catching the exception.

Reply:
> The exception is thrown in case the first byte is a continuation byte because it is simply wrong according to UTF-8 specification.

I agree.

> It caught more than one bug in our UTF-8 handling IIRC so I think it should stay, otherwise bugs will hide more easily.

I think the game should not crush in UTF-8 errors. Because
* players does not like when game is crushing;
* we can not prevent UTF-8 errors -- file can be broke outside the game, inconsistent sequence can be pasted from clipboard and now there is knows bugs.

> The game crashes because nobody catches the exception so if we want to prevent the crash then it should be caught in appropriate places where the function is used directly or indirectly.

Today I will try to implement it.

Reply:
Your condition is incorrect. The original condition checks whether c is the continuation byte which is defined as a  byte
     which starts with 10. If c starts with bits 11 then c & 0x80 would result in 11000000 which is definitely not 0 so your
     condition is true but c is not a continuation byte in this case.

If c starts with bits 11 It would be cathed by previous if condition.

    if((c & 0xE0) == 0xC0)
        return 2;

There are many ways to solve the same problem. Which one is correct is a matter of taste.

Reply:
> If c starts with bits 11 It would be cathed by previous if condition.

Ok, so I went too far as to calling your suggestion incorrect because in our case I guess it would work. However, it relies on the earlier conditions. On the other hand, the current condition is more general and can (should) be easily replaced with a call to `isUtf8ContinuationByte(c)`. The whole if statement could even be moved to the beginning of the function and it would still work, which would not be possible with your suggestion.

> There are many ways to solve the same problem.

Yes.

> Which one is correct is a matter of taste.

No.

You mixed up correctness with form. A solution to a problem is either correct or not, and it only depends on the problem definition. It is not up to taste, correctness can be objectively verified by technical/scientific analysis, or even automatically with, for example, unit tests. It is the form of the solution that is a matter of taste, not correctness.

By the way, I've just realized this code probably has no unit tests. No wonder it causes so many problems.